syntax = "proto3";
package omnistream;

// Represents a single 60Hz tick from the vehicle
message TelemetryPacket {
    string vehicle_id = 1;
    int64 timestamp = 2; // Unix Epoch in Microseconds

    // Simulated Lidar: A scan line of distances (1024 floats)
    repeated float lidar_scan = 3 [packed=true]; 

    // IMU Data
    message IMU {
        float accel_x = 1;
        float accel_y = 2;
        float accel_z = 3;
    }
    IMU imu_reading = 4;

    // Battery level (0.0 - 100.0)
    float battery_level = 5;
}

// Server acknowledgment for streaming
message ServerAck {
    bool success = 1;
    int64 received_timestamp = 2;
}

// Bi-directional streaming service
service TelemetryStream {
    // Client streams telemetry, server sends acks
    rpc StreamTelemetry(stream TelemetryPacket) returns (stream ServerAck);
}